crates/stackable-versioned-macros/src/attrs/item/mod.rs

```diff
@@ -0,0 +1,448 @@
+    ///   that the version of the deprecated action must be greater than the
+    ///   version of the added action.
+    /// - All `changed` actions must use a greater version than `added` but a
+    ///   lesser version than `deprecated`.
+    fn validate_action_order(&self, item_ident: &impl ItemIdentExt) -> Result<()> {
+        let added_version = self.added.as_ref().map(|a| *a.since);
+        let deprecated_version = self.deprecated.as_ref().map(|d| *d.since);
+
+        // First, validate that the added version is less than the deprecated
+        // version.
+        // NOTE (@Techassi): Is this already covered by the code below?
+        if let (Some(added_version), Some(deprecated_version)) = (added_version, deprecated_version)
+        {
+            if added_version > deprecated_version {
+                return Err(Error::custom(format!(
+                    "cannot marked as `added` in version `{added_version}` while being marked as `deprecated` in an earlier version `{deprecated_version}`"
+                )).with_span(item_ident));
+            }
+        }
+
+        // Now, iterate over all changes and ensure that their versions are
+        // between the added and deprecated version.
+        if !self.changes.iter().all(|r| {
+            added_version.map_or(true, |a| a < *r.since)
+                && deprecated_version.map_or(true, |d| d > *r.since)
+        }) {
+            return Err(Error::custom(
+                "all changes must use versions higher than `added` and lower than `deprecated`",
+            )
+            .with_span(item_ident));
+        }
+
+        Ok(())
+    }
+
+    /// This associated function is called by the top-level validation function
+    /// and validates that items use correct names depending on attached
+    /// actions.
+    ///
+    /// The following naming rules apply:
+    ///
+    /// - Fields or variants marked as deprecated need to include the
+    ///   deprecation prefix in their name. The prefix must not be included for
+    ///   fields or variants which are not deprecated.
+    fn validate_item_name(&self, item_ident: &impl ItemIdentExt) -> Result<()> {
+        let starts_with_deprecated = item_ident.starts_with_deprecated_prefix();
+
+        if self.deprecated.is_some() && !starts_with_deprecated {
+            return Err(Error::custom(format!(
+                "marked as `deprecated` and thus must include the `{deprecated_prefix}` prefix",
+                deprecated_prefix = item_ident.deprecated_prefix()
+            ))
+            .with_span(item_ident));
+        }
+
+        if self.deprecated.is_none() && starts_with_deprecated {
+            return Err(Error::custom(
+                format!("not marked as `deprecated` and thus mustn't include the `{deprecated_prefix}` prefix", deprecated_prefix = item_ident.deprecated_prefix())
+            ).with_span(item_ident));
+        }
+
+        Ok(())
+    }
+
+    fn validate_added_action(&self) -> Result<()> {
```

Comment:
The other functions have doc comments, but this doesn't

Reply:
Added in f267020